### PR TITLE
New Spring controller instrumentation modules

### LINKED ISSUE #1576: Spike - Investigate possible ways to determine what annotations exist on classes which implement interfaces that contain annotations.
The problem with annotating interfaces is that those annotations are lost at runtime on the implementing classes. So when you have a controller that implements an interface, and that interface's methods are annotated with a Spring mapping annotation, we can't pick that up, so we have to default to the method name. See the javadoc for [Annotation Type Inherited](https://docs.oracle.com/javase/6/docs/api/java/lang/annotation/Inherited.html) :
```
Note also that this meta-annotation only causes annotations to be inherited from superclasses; annotations on implemented interfaces have no effect
```
The goal of the spike is to see if there is any way to determine these annotations at runtime, possible through reflection of the target interface(s) to fetch annotations and map those to classes that implement the interface(s).

### PR DESCRIPTION
**Pull request by jtduffy**
### Overview
Resolves #1576 
These instrumentation modules support proper transaction naming (route + HTTP method) of traditional annotated spring controllers as well as controllers that inherit annotations from interfaces, super classes or custom annotations. Check one of the README files in the PR for a full description.

spring-4.3.0: This module provides instrumentation for Spring Controllers utilizing Spring Web-MVC v4.3.0 up to but not including v6.0.0. (v6.0.0 instrumentation is provided by another module).
spring-6.0.0: Identical to spring-4.3.0 except for javax -> jakarta namespace.
spring-webflux-controller-mappings-5.0.0: Same functionality as the above modules except for Webflux rather than WebMVC.

Note that because the new instrumentation can change transaction names, enabling this "enhanced transaction naming" is gated by a configuration flag `enhanced_spring_transaction_naming`, which is `false` by default:
```
class_transformer:
    # Enhanced Spring transaction naming.
    # This feature will name any transaction that originates from a Spring controller after
    # the defined route and HTTP method. For example: "/customer/v1/edit (POST)".
    # This includes controllers that implement or extend interfaces/classes with WebMVC related
    # annotations (@RestController, @Controller, @RequestMapping, etc). By default, this is configured
    # to false, which will name transactions for those types of controllers based on the controller
    # class name and method. For example; "CustomerController/edit". This is the naming logic carried
    # over from previous agent versions. "Standard" controllers, with all relevant annotations
    # present on the actual class, will still get named based on route and HTTP method.
    enhanced_spring_transaction_naming: false
```

### Testing
The agent includes a suite of tests which should be used to
verify your changes don't break existing functionality. These tests will run with
Github Actions when a pull request is made. More details on running the tests locally can be found
[here](https://github.com/newrelic/newrelic-java-agent/blob/main/CONTRIBUTING.md),

### Checks

- [x] Your contributions are backwards compatible with relevant frameworks and APIs.
- [x] Your code does not contain any breaking changes. Otherwise please describe. 
- [x] Your code does not introduce any new dependencies. Otherwise please describe.

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -340,6 +341,8 @@ include 'instrumentation:spring-webflux-6.0.0'
 include 'instrumentation:spring-webflux-5.0.0'
 include 'instrumentation:spring-webflux-5.1.0'
 include 'instrumentation:spring-webflux-5.3.0'
+include 'instrumentation:spring-webflux-controller-mappings-5.0.0'
+include 'instrumentation:spring-webclient-5.0'
```
Comment (meiao) [inline settings.gradle:345]: Looks like you re added this webclient that I had removed.

File: instrumentation/spring-6.0.0/build.gradle [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+dependencies {
+    implementation(project(":agent-bridge"))
+    implementation("org.springframework:spring-webmvc:6.0.0")
+    implementation('jakarta.servlet:jakarta.servlet-api:5.0.0')
+}
+
+jar {
+    manifest { attributes 'Implementation-Title': 'com.newrelic.instrumentation.spring-6.0.0',
+            'Implementation-Title-Alias': 'spring_annotations',
+            'Weave-Violation-Filter': 'METHOD_MISSING_REQUIRED_ANNOTATIONS,CLASS_MISSING_REQUIRED_ANNOTATIONS' }
```
Comment (meiao) [inline instrumentation/spring-6.0.0/build.gradle:14]: Same for weave violation filter.

File: instrumentation/spring-6.0.0/src/test/java/com/nr/agent/instrumentation/AbstractHandlerMethodAdapterInstrumentation_withEnhancedNamingFalseTest.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
+package com.nr.agent.instrumentation;
```
Comment (meiao) [inline instrumentation/spring-6.0.0/src/test/java/com/nr/agent/instrumentation/AbstractHandlerMethodAdapterInstrumentation_withEnhancedNamingFalseTest.java:1]: All tests are missing copyright.

File: instrumentation/spring-webflux-controller-mappings-5.0.0/src/main/java/com/nr/agent/instrumentation/web/reactive/SpringControllerUtility.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
+/*
+ *
+ *  * Copyright 2020 New Relic Corporation. All rights reserved.
```
Comment (meiao) [inline instrumentation/spring-webflux-controller-mappings-5.0.0/src/main/java/com/nr/agent/instrumentation/web/reactive/SpringControllerUtility.java:3]: Old copyright

File: instrumentation/spring-6.0.0/src/main/java/com/nr/agent/instrumentation/SpringControllerUtility.java [hunk 5]
```diff
@@ -0,0 +1,274 @@
+/*
+ *
+ *  * Copyright 2020 New Relic Corporation. All rights reserved.
```
Comment (meiao) [inline instrumentation/spring-6.0.0/src/main/java/com/nr/agent/instrumentation/SpringControllerUtility.java:3]: Old Copyright

**Review by meiao — APPROVED** [review]

File: instrumentation/spring-4.3.0/build.gradle [hunk 6]
```diff
@@ -4,9 +4,8 @@ plugins {
...
 dependencies {
     implementation(project(":agent-bridge"))
-    implementation("org.springframework:spring-context:4.3.0.RELEASE")
-    implementation("org.springframework:spring-web:4.3.0.RELEASE")
-    testImplementation("org.jetbrains.kotlin:kotlin-stdlib:1.8.21")
+    implementation("org.springframework:spring-webmvc:4.3.0.RELEASE")
+    implementation('jakarta.servlet:jakarta.servlet-api:4.0.4')
 }
 
 jar {
```
[the remark this answers is not in the document]
Comment (jtduffy) [reply instrumentation/spring-4.3.0/build.gradle:11]: I was going to create another issue to discuss this later.  I don't know if it has any applicability outside of the spring controller stuff or not.

File: instrumentation/spring-4.3.0/src/main/java/com/nr/agent/instrumentation/AbstractHandlerMethodAdapter_Instrumentation.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            HttpServletResponse response, HandlerMethod handlerMethod) throws Exception {
+        Transaction transaction = AgentBridge.getAgent().getTransaction(false);
+
+        if (transaction != null) {
+            Class<?> controllerClass = handlerMethod.getBeanType();
+            Method controllerMethod = handlerMethod.getMethod();
+
+            //If this setting is false, attempt to name transactions the way the legacy point cut
+            //named them
+            boolean isEnhancedNaming = NewRelic.getAgent().getConfig().getValue("class_transformer.enhanced_spring_transaction_naming", false);
```
[the remark this answers is not in the document]
Comment (jtduffy) [reply instrumentation/spring-4.3.0/src/main/java/com/nr/agent/instrumentation/AbstractHandlerMethodAdapter_Instrumentation.java:39]: I'll address this in all 3 modules.

File: instrumentation/spring-4.3.0/src/main/java/com/nr/agent/instrumentation/SpringControllerUtility.java [hunk 8]
```diff
@@ -4,25 +4,226 @@
...
+    public static void assignTransactionNameFromControllerAndMethodRoutes(Transaction transaction, String httpMethod,
+            String rootPath, String methodPath) {
+        httpMethod = httpMethod == null ? "GET" : httpMethod;
+
+        String txnName = getRouteName(rootPath, methodPath, httpMethod);
+        if (NewRelic.getAgent().getLogger().isLoggable(Level.FINEST)) {
+            NewRelic.getAgent()
+                    .getLogger()
+                    .log(Level.FINEST, "SpringControllerUtility::assignTransactionNameFromControllerAndMethodRoutes (6.0.0): calling transaction.setTransactionName to [{0}] " +
+                            "with FRAMEWORK_HIGH and override false, txn {1}, ", txnName, AgentBridge.getAgent().getTransaction().toString());
```
[the remark this answers is not in the document]
Comment (jtduffy) [reply instrumentation/spring-4.3.0/src/main/java/com/nr/agent/instrumentation/SpringControllerUtility.java:159]: I'll address this in all 3 modules.

**Pull request merged**